Trail: Implement GetSearchResultById for custom tag features.

MakeshVineeth opened the pull request:
Moved the code to generate `ISearchResult` in its own function `GenerateSearchResult` Now that function will be called again in `GetSearchResultById` to get a `ValueTask<WikiPreviewSearchResult>`

Involves making two **http** calls: 

1. To get the actual page content by its PageID which is also the SearchObjectId.
2. To get the Bitmap from URL, related code is present in `GenerateSearchResult`

MakeshVineeth committed:
Initial

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +109,88 @@ public ValueTask LoadSearchApplicationAsync()
-                            using var imageClient = new HttpClient();
-                            Stream stream = await imageClient.GetStreamAsync(imgUrl, cancellationToken)
+                            WikiPreviewSearchResult wikiPreviewSearchResult =
+                                await GenerateSearchResult(entry.Value, searchedText).ConfigureAwait(false);
```
Just for your info, the context inside new tasks is the default one, so no really need to use `ConfigureAwait(false)`

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +109,88 @@ public ValueTask LoadSearchApplicationAsync()
         }
 
-        public ValueTask<ISearchResult> GetSearchResultForId(string serializedSearchObjectId)
+        public ValueTask<WikiPreviewSearchResult> GetSearchResultForId(string serializedSearchObjectId)
```
You are not implementing the right method, the right signature of the `interface` is with `object` and not `string`. 
It compiles because in `ISearchApplication` the method has `default` implementation (meaning you don't have to implement it.

MakeshVineeth replied:
Thank you. I've changed it to `object` and casting it. Atlast, returning the `WikiPreviewSearchResult` after await.

MakeshVineeth replied:
This is still not working, kindly let me know on this further.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +109,88 @@ public ValueTask LoadSearchApplicationAsync()
+
+                using var httpClient = new HttpClient();
+                Task<Wiki> task = httpClient.GetFromJsonAsync<Wiki>(url, _serializerOptions);
+                task.Wait();
```
Don't use `Wait` it can potentially cause deadlocks (not in this case), add `async` to this method's signature and use `await`

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +109,88 @@ public ValueTask LoadSearchApplicationAsync()
+                using var httpClient = new HttpClient();
+                Task<Wiki> task = httpClient.GetFromJsonAsync<Wiki>(url, _serializerOptions);
+                task.Wait();
+                PageView pageView = task.Result.Query.Pages.Values.First();
```
You should check here whether `Values` is empty before calling `First`, catching exceptions is relatively heavy and this can cause a little spike while searching.

MakeshVineeth replied:
Checking for the empty using `Count`.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +109,88 @@ public ValueTask LoadSearchApplicationAsync()
+            if (!string.IsNullOrWhiteSpace(resultName))
+            {
+                using var reader = new StringReader(resultName);
+                additionalInfo = await reader.ReadLineAsync().ConfigureAwait(false) ?? resultName;
```
Just for your info, there is nothing asynchronous in `StringReader` unless it reads from a file/memory stream.
If you press `F12` on `ReadLineAsync` method you will see that it simply do `Task.FromResult(ReadLine())` which means, that in this case it allocates a `Task` for no reason.
I suggest you changing to `ReadLine` method.

MakeshVineeth replied:
Yeah, I had this doubt. Thanks for auto-clarification :)

MakeshVineeth committed:
Try implementation & other fixes.

MakeshVineeth committed:
Replace Spaces in URLs with Underscores.

While Wikipedia auto converts this, I've changed it in the code itself to get original URLs for consistency.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -57,14 +58,16 @@ public ValueTask<IHandleResult> HandleSearchResult(ISearchResult searchResult)
             if (string.IsNullOrWhiteSpace(displayedName))
                 return new ValueTask<IHandleResult>(new HandleResult(true, false));
 
+            string titleUrl = displayedName.Replace(' ', '_');
```
Cast the `ISearchResult` to `WikiSearchResult` and use the `URL` property instead of trying to change the `DisplayedName`

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +111,84 @@ public ValueTask LoadSearchApplicationAsync()
+            if (!string.IsNullOrWhiteSpace(resultName))
+            {
+                using var reader = new StringReader(resultName);
+                additionalInfo = reader.ReadLine() ?? resultName;
```
Don't use the same text for `AdditionalInfo` and `ResultName`, it will look weird. If the `ReadLine` results in null or empty data, put null in `AdditionalInfo`

MakeshVineeth replied:
Yeah keeping the repetitive content is not looking good. 

Shall I remove the `AdditionalInfo` all together? It looks more clean that way. As of now, I can't think of any other unique info to show.

adirh3 replied:
Need to find a way to differentiate between results with very similar name, you don't have to use `AdditionalInfo` you can also append it to the `DisplayedName`

MakeshVineeth replied:
I'm not sure if this is what you wanted but Wikipedia API (and also Plugin) already gives search results with pair of braces if there are similar names. The braces can contain some extra terms to differentiate it.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +111,84 @@ public ValueTask LoadSearchApplicationAsync()
+                bitmapImageResult = new BitmapImageResult(); // create empty if no image source.
+            }
+
+            WikiPreviewSearchResult wikiPreviewSearchResult = new()
```
Just `return` the object instead of keeping it in a variable

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +111,84 @@ public ValueTask LoadSearchApplicationAsync()
+            string displayedName = value.Title;
+            double score = displayedName.SearchDistanceScore(searchedText);
+            string pageId = value.PageId.ToString();
+            string wikiUrl = WikiRootUrl + displayedName.Replace(' ','_');
```
You sure this is always a valid URL? Don't you get a valid ID/URL through the API call?
If not, you might want to use `HttpUtility.UrlEncode` instead of trying to fix the string yourself.

MakeshVineeth replied:
I can't get the URL from API but yeah, it is always valid. I'll try the `HttpUtility.UrlEncode`. Thank you.

MakeshVineeth replied:
The method `HttpUtility.UrlEncode` is only giving the undesired results. Since the URLs will always be valid and underscores give the original look of the Wikipedia URLs, I don't think there is any need for it.

adirh3 replied:
Did you use the encode only on the `displayedName`?

MakeshVineeth replied:
Yeah, but it should only keep underscores where the `HttpUtility.UrlEncode` replacing spaces with plus (+) signs & the URL doesn't work with that.

MakeshVineeth committed:
Change the way URLs are generated. Use StringBuilder for Concatenations.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -63,17 +65,19 @@ public ValueTask<IHandleResult> HandleSearchResult(ISearchResult searchResult)
-                    ActionType.Wikipedia => WikiRootUrl + displayedName,
-                    ActionType.Wikiwand => WikiWandUrl + displayedName,
-                    ActionType.GoogleSearch => GoogleSearchUrl + displayedName,
+                    ActionType.Wikipedia => new StringBuilder(WikiRootUrl).Append(wikiPreviewSearchResult.Url)
```
Don't use `StringBuilder` to append a single string, the memory allocation won't be worth it in terms of performance.
Also, the code is less readable this way.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +112,74 @@ public ValueTask LoadSearchApplicationAsync()
-                            Stream stream = await imageClient.GetStreamAsync(imgUrl, cancellationToken)
+                            WikiPreviewSearchResult wikiPreviewSearchResult =
+                                await GenerateSearchResult(entry.Value, searchedText);
+                            await channel.Writer.WriteAsync(wikiPreviewSearchResult, CancellationToken.None)
```
You should check that this is not `null` before writing to channel, since you're returning `default` in some cases in `GenerateSearchResult`

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +112,74 @@ public ValueTask LoadSearchApplicationAsync()
+            }
+            else
+            {
+                bitmapImageResult = new BitmapImageResult(); // create empty if no image source.
```
Why create empty? Just use `null`, FS will put there a question mark icon

MakeshVineeth committed:
Fixes. Change back to default concatenations.

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +110,77 @@ public ValueTask LoadSearchApplicationAsync()
+            string displayedName = value.Title;
+            double score = displayedName.SearchDistanceScore(searchedText);
+            string pageId = value.PageId.ToString();
+            string wikiUrl = new StringBuilder(displayedName).Replace(' ', '_').ToString();
```
Remove the `StringBuilder` usage

adirh3 commented on WikiPreview.Fluent.Plugin/WikiPreviewSearchApp.cs:
```diff
@@ -108,59 +110,77 @@ public ValueTask LoadSearchApplicationAsync()
         }
 
-        public ValueTask<ISearchResult> GetSearchResultForId(string serializedSearchObjectId)
+        public async ValueTask<WikiPreviewSearchResult> GetSearchResultForId(object serializedSearchObjectId)
```
Rename this to `searchObjectId` the serialized part is wrong

MakeshVineeth committed:
Small fixes. Update to 1.1.0.0

MakeshVineeth committed:
Format code and remove redundant toString()
